### ISSUE TITLE
[v2] Overly permissive/redundant IAM permission

### ISSUE THREAD

**Issue by mike-stewart** [kubernetes-sigs/aws-load-balancer-controller#1647]
Related to #1302. Reviewing the required IAM policy for the v2 controller, it appears that there is some duplication.

There are two statements providing "ec2:AuthorizeSecurityGroupIngress" and "ec2:RevokeSecurityGroupIngress". This [more restricted](https://github.com/kubernetes-sigs/aws-load-balancer-controller/blob/842b95e8b0c0ec3bf6b46d0e8d83da2d1b710647/docs/install/iam_policy.json#L97-L110) statement that checks for resource tags, and this [more permissive](https://github.com/kubernetes-sigs/aws-load-balancer-controller/blob/842b95e8b0c0ec3bf6b46d0e8d83da2d1b710647/docs/install/iam_policy.json#L53-L60) statement that does not.

Can the more permissive one be deleted in favour of the more restricted one? Since AuthorizeSecurityGroupIngress is a fairly sensitive permission, it would be great to be able to lock that down further.

**Comment by M00nF1sh** [comment]
@mike-stewart 
we layout the permission like so because they are for different purpose. (we are planning on a doc on how to further scope down these permissions).
the first statement(without resourceTags) is let controller modify worker node security groups, which can be further scope down by `vpc-id` and `kubernetes.io/cluster/cluster-name: owned/shared` tag.
the second statement(with resourceTag) is let controller modify the securityGroup created for ALB. which can be further scope down with `vpc-id` and `elbv2.k8s.aws/cluster: cluster-name` tag.

**Comment by mike-stewart** [comment]
@M00nF1sh Makes sense! Is there an existing issue for the guide on scoping down the permissions? If so, we could close this issue in favour of that one.

**Comment by M00nF1sh** [comment]
@mike-stewart 
we don't have a guide about how to  further scope down the permissions yet. Which we'll work on soon. will update this issue once we have it :D

**Comment by zhujik** [comment]
I'll just highjack this issue to tell you my findings, because we are migrating to v2 and the iam permissions are a bit different.

There are some iam permissions that are now missing that have been used in the previous v1 version, namely:

* acm:GetCertificate
* ec2:DescribeInstanceStatus
* ec2:ModifyInstanceAttribute
* ec2:ModifyNetworkInterfaceAttribute
* shield:DescribeSubscription
* shield:ListProtections
* waf:GetWebACL

For the migration process, for the time being, we have re-included them in the iam policy to avoid any unforseen consequences. So I would just kindly ask if you left them out intentionally or if the might have been omitted accidentally, because the git history of the iam-policy.json does not indicate the history of the file in regards to the v1 version.

**Comment by kishorj** [comment]
@mike-stewart, please refer to the v2.4 live docs https://kubernetes-sigs.github.io/aws-load-balancer-controller/v2.4/deploy/installation/#setup-iam-role-for-service-accounts for scoping the IAM permissions.

**Comment by mike-stewart** [comment]
@kishorj @M00nF1sh Thanks for updating the documentation for the AuthorizeSecurityGroupIngress policy I originally asked about.

I have a few more questions about how to scope down this policy, in particular:
- Could this have a "aws:ResourceTag/elbv2.k8s.aws/cluster" condition? https://github.com/kubernetes-sigs/aws-load-balancer-controller/blob/9c79e45dd4b8cf94b97a1c6737bf1d47306460d0/docs/install/iam_policy.json#L139-L148 
- Could this have a "aws:RequestTag/elbv2.k8s.aws/cluster" condition? https://github.com/kubernetes-sigs/aws-load-balancer-controller/blob/9c79e45dd4b8cf94b97a1c6737bf1d47306460d0/docs/install/iam_policy.json#L167-L179
- Could this have a "aws:ResourceTag/elbv2.k8s.aws/cluster" condition? https://github.com/kubernetes-sigs/aws-load-balancer-controller/blob/9c79e45dd4b8cf94b97a1c6737bf1d47306460d0/docs/install/iam_policy.json#L199-L206
- Could this have a "aws:ResourceTag/elbv2.k8s.aws/cluster" condition? https://github.com/kubernetes-sigs/aws-load-balancer-controller/blob/9c79e45dd4b8cf94b97a1c6737bf1d47306460d0/docs/install/iam_policy.json#L207-L217
- I understand that this one has documentation on how to restrict it further, but why doesn't it have a `"Null": {"aws:ResourceTag/elbv2.k8s.aws/cluster": "false"}` condition to make it more secure by default? https://github.com/kubernetes-sigs/aws-load-balancer-controller/blob/9c79e45dd4b8cf94b97a1c6737bf1d47306460d0/docs/install/iam_policy.json#L68-L75

I think it would be great if the default policy packaged with this app could follow the principle of least privilege to the extent possible without affecting the out-of-the-box install experience of users. If you could broadly apply the "aws:ResourceTag" Null check conditions in the default policy, that would give users an easy path to change those tag conditions from Null check to StringEquals to really lock this down. As it stands, it's not clear if those can be safely scoped down without thorough testing.

/remove-lifecycle stale

**Comment by kpersohn** [comment]
I'm also curious if it's safe to scope these to specific clusters and if so why not just make that the default? I'm more than happy to submit a PR if it would be a welcome contribution.

I saw the null condition check for `"aws:ResourceTag/kubernetes.io/cluster/CLUSTER-NAME": "false"` recommended in the docs for adding/removing rules to security groups and just applied that everywhere there were conditions on `aws:ResourceTag/elbv2.k8s.aws/cluster`. In basic testing, this seems to work fine but not sure if there's a gotcha I haven't encountered yet. 

I agree with @mike-stewart it would be great if the out of box policy was scoped to ownership based on specific cluster name either via StringLike on the value of the generic cluster tag or null conditions on the standard tag where the cluster name is in the key (don't think it really matters which). 

For example: 
```       
       {
            "Effect": "Allow",
            "Action": [
                "elasticloadbalancing:AddTags",
                "elasticloadbalancing:RemoveTags"
            ],
            "Resource": [
                "arn:aws:elasticloadbalancing:*:*:targetgroup/*/*",
                "arn:aws:elasticloadbalancing:*:*:loadbalancer/net/*/*",
                "arn:aws:elasticloadbalancing:*:*:loadbalancer/app/*/*"
            ],
            "Condition": {
                "Null": {
                    "aws:RequestTag/elbv2.k8s.aws/cluster": "true",
                    "aws:ResourceTag/elbv2.k8s.aws/cluster": "false"
                    "aws:RequestTag/kubernetes.io/cluster/CLUSTER-NAME": "true",
                    "aws:ResourceTag/kubernetes.io/cluster/CLUSTER-NAME": "false"
                }
            }
        },
```

**Comment by johngmyers** [comment]
Instead of documentation, perhaps we could provide a tool which takes a cluster name and generates a policy scoped to that cluster?

**Comment by grounded042** [comment]
👋 I agree with the sentiment here. It would be great if better direction was given as to the IAM conditions. Not owning the code it's hard to add in conditions and know if they'll eventually break something or not.

**Comment by mb-shorai** [comment]
I guess the issue is still relevant and should be reopened, there is no information regarding restricting these overly permissive IAM permissions except for the single `ec2:RevokeSecurityGroupIngress`/`ec2:AuthorizeSecurityGroupIngress` statement.

cc @M00nF1sh @kishorj
